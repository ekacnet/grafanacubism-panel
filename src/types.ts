import { DashboardLink } from '@grafana/schema';

export interface CubismOptions {
  text: string;
  extentMin: number;
  extentMax: number;
  automaticExtents: boolean;
  links: DashboardLink[];
}
