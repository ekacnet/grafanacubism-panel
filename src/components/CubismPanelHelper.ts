import { CubismOptions } from 'types';
import * as cubism from 'cubism-es';
import * as d3 from 'd3';

import { PanelData, DataFrame } from '@grafana/data';
import { DataHoverEvent, EventBus } from '@grafana/runtime';
import { getSerieByName, convertAllDataToCubism } from '../cubism_utils';
import { log_debug } from '../misc_utils';
import { calculateSecondOffset } from '../date_utils';

type CSS = string;

interface CSSStyles {
  [key: string]: CSS;
}

export const D3GraphRender = (
  context: cubism.Context,
  data: PanelData,
  options: CubismOptions,
  styles: CSSStyles,
  eventBus: EventBus,
  convertDatahelper: (d: DataFrame[], n: number[], o: any, z: number) => cubism.Metric[] = convertAllDataToCubism
): ((wrapperDiv: HTMLDivElement | null) => void) => {
  return (panelDiv: HTMLDivElement | null) => {
    if (!panelDiv) {
      return;
    }
    if (data.series.length === 0) {
      panelDiv.innerHTML = 'The series contained no data, check your query';
      return;
    }
    // Initialize most of the variables and constants
    let showText = false;

    if (options.text !== undefined && options.text !== null && options.text !== '') {
      showText = true;
    }
    log_debug('Show text is ', showText);
    let now = Date.now();
    let begin = new Date();
    const request = data.request!;

    const start = +request.range.from;
    const end = +request.range.to;
    const span = end - start;

    const anHour = 60 * 60 * 1000;
    const aDay = 24 * anHour;
    const aWeek = 7 * aDay;
    const aMonth = 4 * aWeek;

    // the width of the div with the panel
    let size = panelDiv.clientWidth;

    let step = Math.floor(span / size);

    let cubismTimestamps: number[] = [];

    for (let ts = start; ts <= end; ts = ts + step) {
      cubismTimestamps.push(ts);
    }
    log_debug('Step is:', step);
    log_debug('Length of timestamps is ', cubismTimestamps.length);
    log_debug('Size of the graph is ', size);

    panelDiv.innerHTML = '';
    panelDiv.className = styles['cubism-panel'];
    let cubismData = convertDatahelper(data.series, cubismTimestamps, context, step);

    cubismData = cubismData.filter(function (el) {
      if (el !== null) {
        return true;
      } else {
        return false;
      }
    });

    let prev = +now;
    now = Date.now();
    log_debug(`Took ${now - prev} to convert the series`);

    if (cubismData.length === 0) {
      panelDiv.innerHTML = 'The series contained no data, check your query';
      return;
    }

    // setup Div layout and set classes
    const delta = calculateSecondOffset(begin, +end, request.timezone, request.range.from.utcOffset());
    const cubismGraphDiv = d3.create('div');
    cubismGraphDiv.node()!.className = styles['cubismgraph'];

    const axisDiv = d3.create('div');
    const canvasDiv = d3.create('div');
    canvasDiv.node()!.className = styles['canvas'];

    // setup the context
    // size is the nubmer of pixel
    // steps seems to be the number of microseconds between change
    // it also control the range (ie. given the number of pixel and that a
    // pixel represent 1e3 milliseconds, the range is 1e3 * size seconds)
    context.size(size).step(step);
    context.serverDelay(delta * 1000);
    // negative delta means that we go in the past ...
    context.stop();
    context.setCSSClass('horizon', styles['horizon']);
    context.setCSSClass('line', styles['line']);
    context.setCSSClass('title', styles['title']);
    context.setCSSClass('value', styles['value']);

    // create axis: try to find divs with .axis class
    axisDiv
      .selectAll('.' + styles['axis'])
      .data(['bottom'])
      .enter()
      .append('div')
      .attr('class', styles['axis'])
      .each(function (dataValue) {
        let scale = d3.timeHour;
        let count = 6;
        if (span < 2 * anHour) {
          scale = d3.timeMinute;
          count = 15;
        } else if (span < 12 * anHour) {
          scale = d3.timeHour;
          count = 1;
        } else if (span < aDay) {
          scale = d3.timeHour;
          count = 3;
        } else if (span < 2 * aDay) {
          scale = d3.timeHour;
          count = 6;
        } else if (span < 2 * aWeek) {
          scale = d3.timeDay;
          count = 1;
        } else if (span < 2 * aMonth) {
          scale = d3.timeDay;
          count = 2;
        }
        context.axis().ticks(scale, count).orient(dataValue).render(d3.select(this));
        d3.select(this).attr('class', `${styles['axis']} bottom`);
      });
    prev = now;
    now = Date.now();
    log_debug(`Took ${now - prev} ms to do the axis`);

    // create the horizon
    const h = canvasDiv
      .selectAll(styles['horizon'])
      .data(cubismData)
      .enter()
      .insert('div', '.bottom')
      .attr('class', styles['horizon']);

    // create the rule
    const ruleDiv = canvasDiv.append('div').attr('class', `${styles['rule']}`).attr('id', 'rule');
    const zoomDiv = canvasDiv.append('div').attr('class', 'zoom');
    context.rule().render(ruleDiv);
    context.zoom(zoomCallbackGen(context, data, options)).render(zoomDiv);
    context.zoom().setZoomType('onelane');

    // extent is the vertical range for the values for a given horinzon
    if (options.automaticExtents || options.extentMin === undefined || options.extentMax === undefined) {
      context.horizon().render(h);
    } else {
      context.horizon().extent([options.extentMin, options.extentMax]).render(h);
    }

    context.on('focus', function (i: number) {
      if (i === null) {
        canvasDiv.selectAll('.value').style('right', null);
      } else {
        let val = context._scale.invert(i);
        eventBus.publish(
          new DataHoverEvent({
            point: {
              time: val,
            },
          })
        );
        const rightStyle: string = context.size() - i + 'px';
        canvasDiv.selectAll('.value').style('right', rightStyle);
      }
    });

    cubismGraphDiv.node()!.append(canvasDiv.node()!);
    cubismGraphDiv.node()!.append(axisDiv.node()!);
    panelDiv.append(cubismGraphDiv.node()!);

    if (options.text !== undefined && options.text !== null && options.text !== '') {
      log_debug('showing text');
      let msg = `${options.text}`;
      const msgDivContainer = d3.create('div');
      msgDivContainer.node()!.className = styles['textBox'];
      msgDivContainer.append('div').text(msg);
      panelDiv.append(msgDivContainer.node()!);
    }
    prev = now;
    now = Date.now();
    log_debug(`Took ${now - prev} ms to finish`);
    return;
  };
};

function isString(value: unknown): asserts value is string {
  if (typeof value !== 'string') {
    throw new Error('Not a string');
  }
}

export const zoomCallbackGen = (
  context: cubism.Context,
  data: PanelData,
  options: CubismOptions
): cubism.zoomCallback => {
  const f = (start: number, end: number, selection: cubism.d3Selection) => {
    if (options.links.length === 0) {
      log_debug("Can't do any zoom, there is no links to zoom to");
      return;
    }
    log_debug(`Doing a zoom from point ${start} to point ${end}`);

    if (options.links.length > 1) {
      log_debug('There is more than one link, linked to this graph, will pick the first one');
    }
    let link = options.links[0].url;
    isString(link);
    const titleElement = selection.select('.' + context.getCSSClass('title'));
    let serieName = titleElement.text();
    let field = getSerieByName(data.series, serieName);
    if (field === null) {
      throw new Error(`unable to find the field in the time series associated with the name ${serieName}`);
    }
    let filteredString = link.replace(/\ufeff/g, '');

    const pattern = /\${[^}]*}/g;
    const extractedVariables: string[] = filteredString.match(pattern) || [];
    let unsupported: string[] = [];
    const prefix = '__field.labels';
    for (let j = 0; j < extractedVariables.length; j++) {
      if (!extractedVariables[j].slice(2, -1).startsWith(prefix)) {
        unsupported.push(extractedVariables[j].slice(2, -1));
      } else {
        const label = extractedVariables[j].slice(prefix.length + 3, -1);
        const labels = field.labels!;
        let v = labels[label];
        filteredString = filteredString.replace(extractedVariables[j], encodeURIComponent(v));
      }
    }
    if (unsupported.length > 0) {
      throw new Error('Unsupported variable in the url:' + unsupported.join(', '));
    }
    filteredString = filteredString + '&from=' + +context._scale.invert(start);
    filteredString = filteredString + '&to=' + +context._scale.invert(end);
    if (options.links[0].targetBlank) {
      window.open(filteredString, '_blank');
    } else {
      window.location.assign(filteredString);
    }
  };
  return f;
};
