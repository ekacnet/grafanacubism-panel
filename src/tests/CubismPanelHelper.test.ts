import { D3GraphRender, zoomCallbackGen } from '../components/CubismPanelHelper';
import { CubismOptions } from '../types';
import { convertAllDataToCubism } from '../cubism_utils';
import * as d3 from 'd3';
import * as cubism from 'cubism-es';

import { DashboardLink, DashboardLinkType } from '@grafana/schema';
import { dateTime, toDataFrame, LoadingState, PanelData } from '@grafana/data';

const getValidSerie = (width: number, seed: number, span: number) => {
  const timestamp = 1598919367000;
  const timestamp2 = 1598919367000 + span * 1000;
  let step = (span * 1000) / width;
  let value = 0,
    i = 0;
  let datapoints = [];
  for (let v = timestamp; v <= timestamp2; v += step) {
    datapoints.push([
      (value = Math.max(-10, Math.min(10, value + 0.8 * Math.random() - 0.4 + 0.2 * Math.cos((i += seed * 0.02))))),
      v,
    ]);
  }
  const input1 = {
    target: 'Field Name',
    datapoints: datapoints,
  };
  let series = toDataFrame(input1);

  return series;
};

const getData = (span: number): PanelData => {
  const input1 = {
    target: 'Field Name',
    datapoints: [
      [100, 1],
      [200, 4],
      [1000, 10],
    ],
  };
  const timestamp = 1598919367000;
  const timestamp2 = timestamp + span * 1000;
  const from = dateTime(timestamp);
  const to = dateTime(timestamp2);
  let series = toDataFrame(input1);
  series.fields[1].labels = { instance: 'grafana' };
  let data: PanelData = {
    request: {
      startTime: 12,
      intervalMs: 10,
      app: 'dashboard', // Example field
      requestId: 'Q101', // Example field
      timezone: 'browser', // Example field
      panelId: 1, // Example field
      range: {
        from: from,
        to: to,
        raw: {
          from: 'now-1h', // Example field
          to: 'now', // Example field
        },
      },
      targets: [
        {
          refId: 'A',
        },
      ],
      scopedVars: {}, // Example field
      maxDataPoints: 500, // Example field
      interval: '60s', // Example field
    },
    series: [series],
    state: LoadingState.Done,
    timeRange: {
      from: from,
      to: to,
      raw: {
        from: 'now-1m',
        to: 'now',
      },
    },
  };
  return data;
};

const createMockOptions = (url?: string): CubismOptions => {
  const type: DashboardLinkType = 'link';
  let dashLink: DashboardLink[] = [];
  if (url !== undefined) {
    dashLink.push({
      asDropdown: false,
      icon: '',
      includeVars: false,
      keepTime: false,
      tags: [],
      targetBlank: true,
      title: 'title',
      tooltip: 'tooltip',
      url: url,
      type: type,
    });
  }
  return {
    links: dashLink,
    text: 'foo',
    extentMin: 0,
    extentMax: 10,
    automaticExtents: true,
  };
};

describe('D3GraphRender', () => {
  const width = 300;
  let mockContext: any; // mock the cubism Context
  let mockData: any;
  let mockOptions: any;
  let mockStyles: any;
  let mockPanelDiv: HTMLDivElement;
  let mockTicks: any;
  let mockExtent: any;
  let mockEventBus: any;
  const oldConsole = global.console.log;

  beforeEach(() => {
    mockEventBus = {
      publish: jest.fn(() => {}),
    };
    mockTicks = jest.fn(() => {
      return mockContext.axis();
    });
    mockExtent = jest.fn(() => {
      return mockContext.horizon();
    });
    mockContext = {
      size: jest.fn().mockReturnThis(),
      step: jest.fn().mockReturnThis(),
      serverDelay: jest.fn().mockReturnThis(),
      stop: jest.fn().mockReturnThis(),
      // @ts-ignore
      metric: function (f, n) {
        return this._fake.metric(f, n);
      },
      setCSSClass: jest.fn().mockReturnThis(),
      axis: jest.fn(() => ({
        ticks: mockTicks,
        orient: jest.fn().mockReturnThis(),
        render: jest.fn(),
      })),
      horizon: jest.fn(() => ({
        render: jest.fn(),
        extent: mockExtent,
      })),
      rule: jest.fn(() => ({ render: jest.fn() })),
      zoom: jest.fn(() => ({
        render: jest.fn(),
        setZoomType: jest.fn().mockReturnThis(),
      })),
      on: jest.fn(),
      _fake: cubism.context(),
    };

    mockData = {
      series: [
        /* mock series data here */
      ],
      request: { range: { from: 123, to: 456 }, timezone: 'UTC' },
    };
    mockOptions = {
      text: 'Hello, World!',
      automaticExtents: false,
      extentMin: 0,
      extentMax: 100,
      automaticSampling: true,
      sampleType: false,
    };
    mockStyles = {
      'cubism-panel': 'panel',
      cubismgraph: 'graph',
      canvas: 'canvas',
      axis: 'axis',
      horizon: 'horizon',
      rule: 'rule',
      value: 'value',
      title: 'title',
      textBox: 'text-box',
    };

    mockPanelDiv = document.createElement('div');
    Object.defineProperty(mockPanelDiv, 'clientWidth', {
      value: width,
      writable: false, // Ensuring the property remains read-only
    });

    // Mock console functions for testing
  });

  afterEach(() => {
    global.console.log = oldConsole;
    jest.clearAllMocks();
  });
  it('should not render if panelDiv is null or data.series is empty', () => {
    const renderFn = D3GraphRender(
      mockContext,
      getData(86400),
      mockOptions,
      mockStyles,
      mockEventBus,
      convertAllDataToCubism
    );
    expect(renderFn(null)).toBeUndefined();
  });

  it('should not render graph when data series is empty ', () => {
    const renderFn = D3GraphRender(
      mockContext,
      mockData,
      mockOptions,
      mockStyles,
      mockEventBus,
      convertAllDataToCubism
    );
    expect(renderFn(mockPanelDiv)).toBeUndefined();
  });
  it('should call convertDataToCubism with Auto', () => {
    let data = getData(86400);
    data.series[0].length = 0;
    const mockHelper = jest.fn(() => [null]);
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, mockHelper);
    renderFn(mockPanelDiv);

    const calls = mockHelper.mock.calls;

    expect(calls.length).toBe(1);
    expect(calls[0].at(0)).toBe(data.series);
    expect(calls[0].at(2)).toStrictEqual(mockContext);
    expect(calls[0].at(3)).toStrictEqual(288000);
    expect(mockPanelDiv.innerHTML).toBe('The series contained no data, check your query');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).not.toHaveBeenCalled();
    expect(mockContext.horizon).not.toHaveBeenCalled();
    expect(mockContext.rule).not.toHaveBeenCalled();
    expect(mockContext.zoom).not.toHaveBeenCalled();
  });
  it('should call convertDataToCubism with Auto', () => {
    let data = getData(86400);
    data.series[0].length = 0;
    const mockHelper = jest.fn(() => []);
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, mockHelper);
    renderFn(mockPanelDiv);

    const calls = mockHelper.mock.calls;

    expect(calls.length).toBe(1);
    expect(calls[0].at(0)).toBe(data.series);
    expect(calls[0].at(2)).toStrictEqual(mockContext);
    expect(calls[0].at(3)).toStrictEqual(288000);
    expect(mockPanelDiv.innerHTML).toBe('The series contained no data, check your query');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).not.toHaveBeenCalled();
    expect(mockContext.horizon).not.toHaveBeenCalled();
    expect(mockContext.rule).not.toHaveBeenCalled();
    expect(mockContext.zoom).not.toHaveBeenCalled();
  });
  it('should call convertDataToCubism with Upsample', () => {
    let data = getData(86400);
    data.series[0].length = 0;
    const mockHelper = jest.fn(() => []);
    mockOptions.automaticSampling = false;
    mockOptions.sampleType = false;
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, mockHelper);
    renderFn(mockPanelDiv);

    const calls = mockHelper.mock.calls;

    expect(calls.length).toBe(1);
    expect(calls[0].at(0)).toBe(data.series);
    expect(calls[0].at(2)).toStrictEqual(mockContext);
    expect(calls[0].at(3)).toStrictEqual(288000);
    expect(mockPanelDiv.innerHTML).toBe('The series contained no data, check your query');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).not.toHaveBeenCalled();
    expect(mockContext.horizon).not.toHaveBeenCalled();
    expect(mockContext.rule).not.toHaveBeenCalled();
    expect(mockContext.zoom).not.toHaveBeenCalled();
  });
  it('should call convertDataToCubism with Downsample', () => {
    let data = getData(86400);
    data.series[0].length = 0;
    const mockHelper = jest.fn(() => []);
    mockOptions.automaticSampling = false;
    mockOptions.sampleType = true;
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, mockHelper);
    renderFn(mockPanelDiv);

    const calls = mockHelper.mock.calls;

    expect(calls.length).toBe(1);
    expect(calls[0].at(0)).toBe(data.series);
    expect(calls[0].at(2)).toStrictEqual(mockContext);
    expect(calls[0].at(3)).toStrictEqual(288000);
    expect(mockPanelDiv.innerHTML).toBe('The series contained no data, check your query');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).not.toHaveBeenCalled();
    expect(mockContext.horizon).not.toHaveBeenCalled();
    expect(mockContext.rule).not.toHaveBeenCalled();
    expect(mockContext.zoom).not.toHaveBeenCalled();
  });
  it('should not render if there is not real data', () => {
    const data = getData(86400);
    data.series = [getValidSerie(width, 1, 86400)];
    mockOptions.automaticSampling = false;
    mockOptions.sampleType = false;
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, convertAllDataToCubism);
    // Create a spy on the function

    renderFn(mockPanelDiv);

    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalledWith(d3.timeHour, 6);
    expect(mockContext.horizon).toHaveBeenCalled();
    expect(mockContext.rule).toHaveBeenCalled();
    expect(mockContext.zoom).toHaveBeenCalled();
  });
  it('should render graph and text when panelDiv and data are valid no extent', () => {
    const data = getData(86400);
    data.series = [getValidSerie(width, 1, 86400)];
    mockOptions.automaticExtents = true;
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, convertAllDataToCubism);
    renderFn(mockPanelDiv);

    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalled();
    expect(mockExtent).not.toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalledWith(d3.timeHour, 6);
    expect(mockContext.horizon).toHaveBeenCalled();
    expect(mockContext.rule).toHaveBeenCalled();
    expect(mockContext.zoom).toHaveBeenCalled();
  });
  it('should render graph and text when panelDiv and data are valid', () => {
    const data = getData(86400);
    data.series = [getValidSerie(width, 1, 86400)];
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, convertAllDataToCubism);
    renderFn(mockPanelDiv);

    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).toHaveBeenCalled();
    expect(mockExtent).toHaveBeenCalledWith([mockOptions.extentMin, mockOptions.extentMax]);
    expect(mockTicks).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalledWith(d3.timeHour, 6);
    expect(mockContext.horizon).toHaveBeenCalled();
    expect(mockContext.rule).toHaveBeenCalled();
    expect(mockContext.zoom).toHaveBeenCalled();
  });
  it('should render graph and text when panelDiv and data are valid for more than a 14 day', () => {
    let time = 14 * 86400;
    const data = getData(time);
    data.series = [getValidSerie(width, 1, time)];
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, convertAllDataToCubism);
    renderFn(mockPanelDiv);

    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalledWith(d3.timeDay, 2);
    expect(mockContext.horizon).toHaveBeenCalled();
    expect(mockContext.rule).toHaveBeenCalled();
    expect(mockContext.zoom).toHaveBeenCalled();
  });
  it('should render graph and text when panelDiv and data are valid for less than a 14 day', () => {
    let time = 14 * 86400 - 1;
    const data = getData(time);
    data.series = [getValidSerie(width, 1, time)];
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, convertAllDataToCubism);
    renderFn(mockPanelDiv);

    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalledWith(d3.timeDay, 1);
    expect(mockContext.horizon).toHaveBeenCalled();
    expect(mockContext.rule).toHaveBeenCalled();
    expect(mockContext.zoom).toHaveBeenCalled();
  });
  it('should render graph and text when panelDiv and data are valid for less than a 1/2 day', () => {
    let time = 86400 / 2 - 1;
    const data = getData(time);
    data.series = [getValidSerie(width, 1, time)];
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, convertAllDataToCubism);
    renderFn(mockPanelDiv);

    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalledWith(d3.timeHour, 1);
    expect(mockContext.horizon).toHaveBeenCalled();
    expect(mockContext.rule).toHaveBeenCalled();
    expect(mockContext.zoom).toHaveBeenCalled();
  });
  it('should render graph and text when panelDiv and data are valid for less than a day', () => {
    let time = 86400 - 1;
    const data = getData(time);
    data.series = [getValidSerie(width, 1, time)];
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, convertAllDataToCubism);
    renderFn(mockPanelDiv);

    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalledWith(d3.timeHour, 3);
    expect(mockContext.horizon).toHaveBeenCalled();
    expect(mockContext.rule).toHaveBeenCalled();
    expect(mockContext.zoom).toHaveBeenCalled();
  });
  it('should render graph and text when panelDiv and data are valid and called for an hour ', () => {
    const data = getData(3500);
    data.series = [getValidSerie(width, 1, 3500)];
    const renderFn = D3GraphRender(mockContext, data, mockOptions, mockStyles, mockEventBus, convertAllDataToCubism);
    renderFn(mockPanelDiv);

    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
    expect(mockContext.axis).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalled();
    expect(mockTicks).toHaveBeenCalledWith(d3.timeMinute, 15);
    expect(mockContext.horizon).toHaveBeenCalled();
    expect(mockContext.rule).toHaveBeenCalled();
    expect(mockContext.zoom).toHaveBeenCalled();
  });
});

describe('focusCallback', () => {
  let context: any;
  let mockOptions: any;
  let mockStyles: any;
  let mockPanelDiv: HTMLDivElement;
  let mockEventBus: any;

  beforeEach(() => {
    context = cubism.context();

    mockEventBus = {
      publish: jest.fn(() => {}),
    };

    mockOptions = {
      text: 'Hello, World!',
      automaticExtents: false,
      extentMin: 0,
      extentMax: 100,
      automaticSampling: true,
      sampleType: false,
    };
    mockStyles = {
      'cubism-panel': 'panel',
      cubismgraph: 'graph',
      canvas: 'canvas',
      axis: 'axis',
      horizon: 'horizon',
      rule: 'rule',
      value: 'value',
      title: 'title',
      textBox: 'text-box',
    };

    mockPanelDiv = document.createElement('div');
    Object.defineProperty(mockPanelDiv, 'clientWidth', {
      value: 300,
      writable: false, // Ensuring the property remains read-only
    });
  });
  it('should call eventBus when context.focus() is called', () => {
    const width = 300;
    document.body.innerHTML = '<div id="demo"></div>';

    mockPanelDiv = document.createElement('div');
    Object.defineProperty(mockPanelDiv, 'clientWidth', {
      value: width,
      writable: false, // Ensuring the property remains read-only
    });

    // Mock console functions for testing
    context = cubism.context();
    let data = getData(3500);
    data.series = [getValidSerie(width, 1, 3500)];
    mockEventBus.publish();
    const renderFn = D3GraphRender(context, data, mockOptions, mockStyles, mockEventBus);

    // @ts-ignore
    renderFn(mockPanelDiv);

    context.focus(12);
    expect(mockEventBus.publish).toHaveBeenCalled();
    expect(mockEventBus.publish).toHaveBeenCalledWith({
      origin: undefined,
      payload: {
        point: { time: new Date('2020-09-01T00:18:11.283Z') },
      },
      type: 'data-hover',
    });
    expect(mockPanelDiv.innerHTML).not.toBe('');
    expect(mockPanelDiv.className).toBe(mockStyles['cubism-panel']);
  });
});

describe('zoomCallbackGen', () => {
  let context: cubism.Context;
  let data: PanelData;
  let options: CubismOptions;

  const getSelection = () => {
    document.body.innerHTML = '<div id="demo"></div>';
    let h = d3.select('#demo');
    return h;
  };
  beforeEach(() => {
    options = createMockOptions('http://example.com');
    context = cubism.context();
    data = getData(86400);
  });
  it('should complain if there is more than one link', () => {
    options = createMockOptions('http://example.com');
    const l = options.links[0];
    options.links.push(l);
    const zoomCallback = zoomCallbackGen(context, data, options);
    const oldFunc = console.log;
    console.log = jest.fn();
    const mockSelection = getSelection();
    const oldFunc2 = global.window.open;
    global.window.open = jest.fn();

    mockSelection.select = jest.fn().mockReturnValueOnce({ text: jest.fn().mockReturnValue('grafana') });

    zoomCallback(0, 100, mockSelection);
    expect(console.log).toHaveBeenCalledWith(
      'There is more than one link, linked to this graph, will pick the first one'
    );
    console.log = oldFunc;
    global.window.open = oldFunc2;
  });
  it('should log a message if there are no links to zoom to', () => {
    options = createMockOptions();
    const zoomCallback = zoomCallbackGen(context, data, options);
    const oldFunc = console.log;
    console.log = jest.fn();

    zoomCallback(0, 100, getSelection());
    expect(console.log).toHaveBeenCalledWith("Can't do any zoom, there is no links to zoom to");
    console.log = oldFunc;
  });
  it('should open a link in a new tab if targetBlank is true', () => {
    options = createMockOptions('http://example.com/&i=${__field.labels.instance}');
    options.links[0].targetBlank = false;
    const zoomCallback = zoomCallbackGen(context, data, options);

    window = Object.create(window);
    const url = 'http://dummy.com';
    Object.defineProperty(window, 'location', {
      value: {
        href: url,
      },
      writable: true, // possibility to override
    });
    window.location.assign = jest.fn();
    const mockSelection = getSelection();

    mockSelection.select = jest.fn().mockReturnValueOnce({ text: jest.fn().mockReturnValue('grafana') });

    zoomCallback(0, 100, mockSelection);
    // @ts-ignore
    const callArgs = global.window.location.assign.mock.calls;
    const urlCalled = callArgs[0][0];

    // Expectation
    expect(urlCalled.startsWith('http://example.com/&i=grafana&from=')).toBe(true);
  });
  it('should open a link in a new tab if targetBlank is true', () => {
    options = createMockOptions('http://example.com/&i=${__field.labels.instance}');
    const zoomCallback = zoomCallbackGen(context, data, options);
    const oldFunc = global.window.open;
    global.window.open = jest.fn();
    const mockSelection = getSelection();

    mockSelection.select = jest.fn().mockReturnValueOnce({ text: jest.fn().mockReturnValue('grafana') });

    zoomCallback(0, 100, mockSelection);
    expect(global.window.open).toHaveBeenCalledWith(expect.any(String), '_blank');
    global.window.open = oldFunc;
  });
  it('should throw an error if field is not found', () => {
    const zoomCallback = zoomCallbackGen(context, data, options);
    const mockSelection = getSelection();

    mockSelection.select = jest.fn().mockReturnValueOnce({ text: jest.fn().mockReturnValue('_field.name') });
    expect(() => zoomCallback(0, 100, mockSelection)).toThrowErrorMatchingSnapshot();
  });
  it('should throw an error if there is no associated label with a variable', () => {
    options = createMockOptions('http://example.com/&i=${__field.labels.foo}');
    data.series[0].fields[1].labels = undefined;
    const zoomCallback = zoomCallbackGen(context, data, options);
    const mockSelection = getSelection();

    mockSelection.select = jest.fn().mockReturnValueOnce({ text: jest.fn().mockReturnValue('grafana') });
    expect(() => zoomCallback(0, 100, mockSelection)).toThrow();
  });
  it('should throw an error if there is no labels', () => {
    options = createMockOptions('http://example.com/&i=${__fields.something.instance}');
    const zoomCallback = zoomCallbackGen(context, data, options);
    const mockSelection = getSelection();

    mockSelection.select = jest.fn().mockReturnValueOnce({ text: jest.fn().mockReturnValue('grafana') });
    expect(() => zoomCallback(0, 100, mockSelection)).toThrow();
  });
  it('should throw an error if url is undefined', () => {
    options = createMockOptions('http://example.com/&i=${__fields.something.instance}');
    options.links[0].url = undefined;
    const zoomCallback = zoomCallbackGen(context, data, options);
    const mockSelection = getSelection();

    mockSelection.select = jest.fn().mockReturnValueOnce({ text: jest.fn().mockReturnValue('grafana') });
    expect(() => zoomCallback(0, 100, mockSelection)).toThrow();
  });
});
