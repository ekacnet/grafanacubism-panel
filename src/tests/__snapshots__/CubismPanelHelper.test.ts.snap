// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zoomCallbackGen should throw an error if field is not found 1`] = `"unable to find the field in the time series associated with the name _field.name"`;
